File: server/src/main/java/com/oceanbase/ocp/controller/perf/ObPlanController.java
/*
 * Copyright (c) 2023 OceanBase
 * OCP Express is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 */

package com.oceanbase.ocp.controller.perf;

import java.time.OffsetDateTime;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.oceanbase.ocp.core.response.IterableResponse;
import com.oceanbase.ocp.core.response.ResponseBuilder;
import com.oceanbase.ocp.core.response.SuccessResponse;
import com.oceanbase.ocp.perf.sql.PlanExplainService;
import com.oceanbase.ocp.perf.sql.PlanRawStatService;
import com.oceanbase.ocp.perf.sql.model.PlanExplain;
import com.oceanbase.ocp.perf.sql.model.PlanStatGroup;
import com.oceanbase.ocp.perf.sql.param.QueryPlanExplainParam;
import com.oceanbase.ocp.perf.sql.param.QueryTopPlanParam;

@RestController
@Validated
@RequestMapping("/api/v1/ob/tenants/{tenantId}")
public class ObPlanController {

    @Autowired
    PlanRawStatService planRawStatService;

    @Autowired
    private PlanExplainService planExplainService;

    @GetMapping(value = "/sqls/{sqlId}/topPlanGroup", produces = {"application/json"})
    public IterableResponse<PlanStatGroup> topPlanGroup(@PathVariable("tenantId") Long tenantId,
            @RequestParam(value = "dbName", required = false) String dbName,
            @PathVariable("sqlId") String sqlId,
            @RequestParam(value = "startTime") OffsetDateTime startTime,
            @RequestParam(value = "endTime") OffsetDateTime endTime) {
        QueryTopPlanParam param = QueryTopPlanParam.builder()
                .startTime(startTime)
                .endTime(endTime)
                .tenantId(tenantId)
                .sqlId(sqlId)
                .dbName(dbName)
                .attachPlanExplain(true)
                .build();
        List<PlanStatGroup> res = planRawStatService.topPlanGroup(param);
        return ResponseBuilder.iterable(res);
    }

    @GetMapping(value = "/plans/{planUid}/explain", produces = {"application/json"})
    public SuccessResponse<PlanExplain> planExplain(
            @PathVariable("tenantId") Long tenantId,
            @PathVariable("planUid") String planUid,
            @RequestParam(value = "startTime") OffsetDateTime startTime,
            @RequestParam(value = "endTime") OffsetDateTime endTime) {
        return ResponseBuilder.single(planExplainService.getExplain(QueryPlanExplainParam.builder()
                .startTime(startTime)
                .endTime(endTime)
                .tenantId(tenantId)
                .planUid(planUid)
                .build()));
    }
}

File: frontend/src/service/ocp-express/ObPlanController.ts
/*
 * Copyright (c) 2023 OceanBase
 * OCP Express is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 */

/* eslint-disable */
// 该文件由 OneAPI 自动生成，请勿手动修改！
import request from '@/util/request';

/** 查询Plan的计划
@param tenantId 租户Id
@param planUid 计划的唯一Id
@param startTime 期间开始时间
@param endTime 期间结束时间
 GET /api/v1/ob/tenants/${param0}/plans/${param1}/explain */
export async function planExplain(
  params: {
    // query
    /** 期间开始时间 */
    startTime?: string;
    /** 期间结束时间 */
    endTime?: string;
    // path
    /** 租户Id */
    tenantId?: number;
    /** 计划的唯一Id */
    planUid?: string;
  },
  options?: { [key: string]: any },
) {
  const { tenantId: param0, planUid: param1, ...queryParams } = params;
  return request<API.SuccessResponse_PlanExplain_>(
    `/api/v1/ob/tenants/${param0}/plans/${param1}/explain`,
    {
      method: 'GET',
      params: {
        ...queryParams,
      },
      ...(options || {}),
    },
  );
}

/** 查询SQL的TopPlan， 按plan hash聚合
@param tenantId 租户Id
@param sqlId SQL_ID
@param startTime 期间开始时间
@param endTime 期间结束时间
@return
 GET /api/v1/ob/tenants/${param0}/sqls/${param1}/topPlanGroup */
export async function topPlanGroup(
  params: {
    // query
    dbName?: string;
    /** 期间开始时间 */
    startTime?: string;
    /** 期间结束时间 */
    endTime?: string;
    // path
    /** 租户Id */
    tenantId?: number;
    /** SQL_ID */
    sqlId?: string;
  },
  options?: { [key: string]: any },
) {
  const { tenantId: param0, sqlId: param1, ...queryParams } = params;
  return request<API.IterableResponse_PlanStatGroup_>(
    `/api/v1/ob/tenants/${param0}/sqls/${param1}/topPlanGroup`,
    {
      method: 'GET',
      params: {
        ...queryParams,
      },
      ...(options || {}),
    },
  );
}
